pkg/controller/machine.go

```diff
@@ -374,99 +334,180 @@ func (c *Controller) syncHandler(key string) error {
+// deleteMachineAndProviderInstance makes sure that an instance has gone in a series of steps.
+func (c *Controller) deleteMachineAndProviderInstance(prov cloud.Provider, providerInstance instance.Instance, machine *machinev1alpha1.Machine) error {
+	var err error
+
+	// step 1: delete provider instance
+	if err = c.deleteProviderInstance(prov, machine); err != nil {
+		if ok, _, _ := cloudprovidererrors.IsTerminalError(err); ok {
+			message := fmt.Sprintf("%v. Please manually delete finalizers from the machine object.", err)
+			if _, errNested := c.updateMachineError(machine, machinev1alpha1.DeleteMachineError, message); errNested != nil {
+				return fmt.Errorf("failed to update machine error after failed delete: %v", errNested)
+			}
+			return err
+		}
+		return fmt.Errorf("failed to delete machine at cloudprovider: %v", err)
+	}
+
+	// step 2: check that the instance has really gone
+	if providerInstance, err = c.getProviderInstance(prov, machine); err == nil {
+		if sets.NewString(string(instance.StatusDeleted), string(instance.StatusDeleting)).Has(string(providerInstance.Status())) {
+			glog.V(4).Infof("deletion of instance %s got triggered. Waiting until it fully disappears", providerInstance.ID())
+			c.workqueue.AddAfter(machine.Name, deletionRetryWaitPeriod)
+			return nil
+		}
+		glog.V(2).Infof("instance %s of machine %s is not deleted yet", providerInstance.ID(), machine.Name)
+		c.workqueue.AddAfter(machine.Name, deletionRetryWaitPeriod)
+		return nil
+	}
+	if err != cloudprovidererrors.ErrInstanceNotFound {
+		if ok, _, _ := cloudprovidererrors.IsTerminalError(err); ok {
+			if _, errNested := c.updateMachineError(machine, machinev1alpha1.DeleteMachineError, err.Error()); errNested != nil {
+				return fmt.Errorf("failed to update machine error after failed delete: %v", errNested)
+			}
+			return err
+		}
+		return fmt.Errorf("failed to get instance for machine %s after the delete got triggered", machine.Name)
+	}
+	glog.V(4).Infof("Deleted machine %s at cloud provider", machine.Name)
+
+	// step 3: remove finalizers this essentially will remove the machine object from the system
+	finalizers := sets.NewString(machine.Finalizers...)
+	finalizers.Delete(finalizerDeleteInstance)
+	machine.Finalizers = finalizers.List()
+	if machine, err = c.updateMachine(machine); err != nil {
+		return fmt.Errorf("failed to update machine after removing the delete instance finalizer: %v", err)
+	}
+	glog.V(4).Infof("Removed delete finalizer from machine %s", machine.Name)
+
+	// step 4: remove error message in case it was set
+	if machine, err = c.clearMachineErrorIfSet(machine); err != nil {
+		return fmt.Errorf("failed to update machine after removing the delete error: %v", err)
+	}
+	return nil
+}
+
+func (c *Controller) createNodeForMachine(prov cloud.Provider, providerInstance instance.Instance, machine *machinev1alpha1.Machine, userdataProvider userdata.Provider, providerConfig *providerconfig.Config) error {
```

Comment:
Maybe rename to something like `createCloudProviderInstanceForMachine` or `createInstanceForMachine` ?
`Node` might be a bit confusing as we have the kubernetes-node resource we handle as well

Reply:
I like `createInstanceForMachine`